### ISSUE TITLE
POST requests that aren't x-www-form-urlencoded?

### ISSUE THREAD

**Issue by magnars** [xeqi/peridot#22]
Hello! Thanks for creating this nice tool. I'm using it to test an API that takes EDN in the post body. I see that the `body` defmulti has support for passing a String and it being served to Ring as a `ByteArrayInputStream`, but the `request` method itself foils it by adding content-type `application/x-www-form-urlencoded` (and the `multipart/multipart?` fails if you don't pass it a map).

Is there any support for POSTing non-urlencoded with peridot? Any thoughts on how to do it?

**Comment by magnars** [comment]
To fix this for myself, I added this:

```clj
(defn handle-edn-body [request]
  (if-let [edn (:edn-body request)]
    (-> request
        (mock/body (pr-str edn))
        (mock/content-type "application/edn"))
    request))
```

and then hijacked `peridot.core/request` and changed

```clj
(let [request (pr/build uri env headers cookie-jar content-type)
```

to

```clj
(let [request (handle-edn-body (pr/build uri env headers cookie-jar content-type))
```

Which seems to be working fine. Maybe support for something along these lines would be interesting to bring into peridot itself?

**Comment by glenjamin** [comment]
I think you should be able to do this by passing the `:content-type` key into the `(request)` function as seen in https://github.com/xeqi/peridot/blob/master/test/peridot/test/core.clj#L83

I'll leave this issue open as we could do with better docs for what `(request)` supports

**Comment by magnars** [comment]
Yes, you can set your own content-type ahead of time, and it won't be overwritten, but `multipart/multipart?` still throws an exception for non-map `:params`

**Comment by glenjamin** [comment]
I think I'm missing something, this test passes: 59f37ae

The only way `multipart/multipart?` would receive non-map `:params` would be if it was passed explicitly in `(request)`.

Can you provide a fuller example?

**Comment by magnars** [comment]
Ah, look at that. I tried posting it as `:params`, not `:body`. I couldn't see the `:body` added to the request anywhere in `peridot.request/build`. I now see it is added along with most everything else in `add-env`. Thanks!

**Comment by mkorvas** [comment]
I know this ticket is closed -- but I have run into the exact problem that the title of this issue implies. I tried to construct and send to the app a POST request with Content-Type: application/json but was getting rejected by the handler with 415 Unsupported Media Type.

Reading the source a few times (I am new to Clojure...) and seeing this discussion, I figured that to send such a request, I have to say not just
```Clojure
(request uri
         :request-method :post
         :content-type "application/json"
         :body some-json-string)
```
(headers ended up being only `{"host" "localhost"}`)
or
```Clojure
(request uri
         :request-method :post
         :headers {"content-type" "application/json"}
         :body some-json-string)
```
(headers got overwritten to `{"host" "localhost", "content-type" "application/x-www-form-urlencoded"}`)
but I **must include both `:headers` and `:content-type`**:
```Clojure
(request uri
         :request-method :post
         :headers {"content-type" "application/json"}
         :content-type "application/json"
         :body some-json-string)
```

I was reluctant to call `(content-type "application/json")` in the "thread" as the docs say it makes the setting persistent also for following requests which I certainly didn't want... although I now realize that specifying `:content-type "application/json"` in the `(request)` arguments probably has the same effect.

This wasn't really clear from the docs, so I would find it nice to mention this feature there. Even nicer would be to be able to ask for removing the `:content-type` key from the session upon issuing the next request to the app -- not having the `:content-type` setting persist for all future requests. (I will be glad if you tell me I missed something and this behaviour is already there!)
